include/zephyr/drivers/espi_emul.h

```diff
@@ -144,42 +126,61 @@ int espi_emul_register(const struct device *dev, const char *name,
+ * Perform the I/O read through the eSPI bus
  *
- * @retval 0 If successful.
- * @retval -EIO General input / output error.
+ * @param dev Device that will handle the operation
+ * @param length number of bytes to read (1/2/4 supported)
+ * @param addr I/O address to read
+ * @param reg Pointer to value which will store the data
+ * @return 0 if IO was successful, -EINVAL if address is not supported
  */
-int emul_espi_host_send_vw(const struct device *espi_dev,
-			   enum espi_vwire_signal vw, uint8_t level);
+int espi_emul_io_read(const struct device *dev,
+	uint8_t length, uint16_t addr, uint32_t *reg);
 
 /**
- * Perform port80 write on the emulated host side, which will
- * trigger a proper event(and callbacks) on the emulated eSPI controller
+ * Perform the I/O write through the eSPI bus
  *
- * @param espi_dev eSPI emulation controller device
- * @param data The date to be written.
+ * @param dev Device that will handle the operation
+ * @param length number of bytes to write (1/2/4 supported)
+ * @param addr I/O address to write
+ * @param reg Data to write
+ * @return 0 if IO was successful, -EINVAL if address is not supported
+ */
+int espi_emul_io_write(const struct device *dev,
+	uint8_t length, uint16_t addr, uint32_t reg);
+
+/**
+ * Set the internal value of virtual-wire on the host side.
  *
- * @retval 0 If successful.
- * @retval -EIO General input / output error.
+ * @param dev Device that will handle the operation
+ * @param vw virtual wire which value will be changed
+ * @param level logic level on virtual wire
+ * @return 0 if IO was successful, -EIO if there's no emulator, -EPERM if
+ *           the direction of vwire is invalid
  */
-int emul_espi_host_port80_write(const struct device *espi_dev, uint32_t data);
+int espi_emul_internal_set_vwire(const struct device *dev,
+	enum espi_vwire_signal vw, uint8_t level);
 
-#ifdef CONFIG_ESPI_PERIPHERAL_ACPI_SHM_REGION
 /**
- * Get the host device's ACPI shared memory start address. The size of the region is
- * CONFIG_EMUL_ESPI_HOST_ACPI_SHM_REGION_SIZE.
+ * Manage the callback on the host side.
  *
- * @param espi_dev eSPI emulation controller device.
- * @return Address of the start of the ACPI shared memory.
+ * @param dev Device that will handle the operation
+ * @param callback structure describing managed callback
+ * @param set enable or disable the callback
+ * @return 0 if IO was successful, -EIO if there's no emulator, or other values
+ *           returned by the emulator
  */
-uintptr_t emul_espi_host_get_acpi_shm(const struct device *espi_dev);
-#endif
+int espi_emul_manage_host_callback(const struct device *dev,
+	struct espi_callback *callback, bool set);
 
+/**
+ * Call the callbacks registered on the host side.
+ *
+ * @param dev Device that will handle the operation
+ * @param ev structure describing event about which the host will be informed
+ * @return always 0
+ */
+int espi_emul_raise_event(const struct device *dev, struct espi_event ev);
```

Comment:
Is this something that the emulator would be expected to call spontaneously, or will it be called from outside the emulator? Should it be part of the API?

Reply:
This function is used only to trigger the callbacks for tests. In normal way, these callbacks are triggered by the interrupts from the MCU, but in tests they have to be called manually from the tests.
Its usage designed to be used by e.g. host emulator, which calls it in response to internal change of vwires, simulating the process of sending the vwire change through eSPI bus to the MCU.

Reply:
It might be helpful to distinguish in the API, either with separate structs or just comments, which functions are expected to be called from the EC (or whatever the generic name for the UUT would be in this context) and which are are expected to be called from the (emulated) host.